### PR TITLE
Optimize Search

### PR DESCRIPTION
**Pull request by VishnuSanal**
## Description

- [x] store recent search items in `SharedPreferences`
- [x] search in the current folder first
- [x] search in the `MediaStore` next
- [x] search in the whole device, if prompted by the user later

#### Automatic tests
- [ ] Added test cases
  
#### Manual tests
- [x] Done  
- Device: Realme Narzo 20A
- OS: Android 11
- Device: Pixel 3A Emulator
- OS: Android 13

#### Build tasks success  
Successfully running following tasks on local:
- [x] `./gradlew assembledebug`
- [x] `./gradlew spotlessCheck`

### CODE REVIEW

**Comment by VishnuSanal** [comment]
Build test APK

**Comment by VishalNehra** [comment]
> The requested APKs has been built. Please find them from the artifacts section of this PR.

Since when do we have this 😮

**Comment by VishnuSanal** [comment]
> > The requested APKs has been built. Please find them from the artifacts section of this PR.
> 
> Since when do we have this 😮 Dude even did a thumbs up emoji

🥺 #3647

**Comment by VishnuSanal** [comment]
@VishnuSanal BTW: I've also tried to link to the artifacts, but seems like it's not as easy as it seems - please see https://github.com/TeamAmaze/AmazeFileManager/compare/hotfix/3.8.6...update-ondemand-workflow

> runs, but fetches artifacts from the previous run. seems like the API responses are updated _after_ a workflow. then, it is not possible to fetch the artifact links from the current workflow from itself.

**Comment by VishnuSanal** [comment]
update: ready for review 🥳 

TODO:

- [ ] move all search operations to BG thread; currently freezes the UI
- [ ] improve new UI ;)

File: app/src/main/java/com/amaze/filemanager/ui/activities/MainActivityViewModel.kt [hunk 1]
```diff
@@ -72,4 +82,96 @@ class MainActivityViewModel(val applicationContext: Application) :
...
+            PreferenceManager.getDefaultSharedPreferences(mainActivity)
+                .getBoolean(PREFERENCE_SHOW_HIDDENFILES, false)
+
+        viewModelScope.launch(Dispatchers.IO) {
+            val projection = arrayOf(MediaStore.Files.FileColumns.DATA)
+
+            val cursor = mainActivity
+                .contentResolver
+                .query(MediaStore.Files.getContentUri("external"), projection, null, null, null)
+                ?: return@launch
```
Comment (VishnuSanal) [inline app/src/main/java/com/amaze/filemanager/ui/activities/MainActivityViewModel.kt:146]: @VishalNehra please disable "Expression with labels increase complexity and affect maintainability." on Codacy. In the case of this PR, I can't remove that label, or else the code wouldn't compile!

**Comment by VishnuSanal** [comment]
Build test APK

**Review by VishalNehra — APPROVED** [review]

**Comment by VishalNehra** [comment]
Amazing work here!

**Comment by VishnuSanal** [comment]
> Amazing work here!

Thank you!

**Pull request merged**